### PR TITLE
Properly save and restore dynamic tab state

### PR DESCRIPTION
**Pull request by kburtram**
This fixes a bug when tab switching where all documents share the state of the most recently added dynamic tab.

### CODE REVIEW

File: src/sql/workbench/parts/query/modelViewTab/queryModelViewTab.ts [hunk 1]
```diff
@@ -28,9 +28,14 @@ export class QueryModelViewTab implements IPanelTab {
 	}
 }
 
-export class QueryModelViewTabView implements IPanelView {
+export class QueryModelViewState {
+	public componentId: string;
```
Comment (udeeshagautam) [inline src/sql/workbench/parts/query/modelViewTab/queryModelViewTab.ts]: why do we have this and getter setter with same name to this.state.componentId?

File: src/sql/workbench/parts/query/common/queryResultsInput.ts [hunk 2]
```diff
@@ -12,21 +12,48 @@ import { MessagePanelState } from 'sql/workbench/parts/query/browser/messagePane
...
 	public messagePanelState: MessagePanelState = new MessagePanelState();
 	public chartState: ChartState = new ChartState();
 	public queryPlanState: QueryPlanState = new QueryPlanState();
 	public topOperationsState = new TopOperationsState();
+	public dynamicModelViewTabsState: Map<string, QueryModelViewState> = new Map<string, QueryModelViewState>();
+
 	public activeTab: string;
 	public visibleTabs: Set<string> = new Set<string>();
 
+	public saveDynamicTabState(tabs: QueryModelViewTab[]): void {
```
Comment (anthonydresser) [inline src/sql/workbench/parts/query/common/queryResultsInput.ts]: This relationship should be flipped, the results view state should be inserted into the tabs and the tabs should be responsible for restoreing their own state.

Comment (kburtram) [reply src/sql/workbench/parts/query/common/queryResultsInput.ts]: @anthonydresser  could you clarify specifically what you'd like to see changed?  You think restoreDynamicTabState should be moved into the tab class?

Comment (anthonydresser) [reply src/sql/workbench/parts/query/common/queryResultsInput.ts]: The QueryModelViewTab should have a state setting method that takes in the result view state object and restores it's own state.

Comment (kburtram) [reply src/sql/workbench/parts/query/common/queryResultsInput.ts]: sure, let me take a look

Comment (kburtram) [reply src/sql/workbench/parts/query/common/queryResultsInput.ts]: @anthonydresser I've move the state logic into QueryModelViewTab.  Please take another look.

File: src/sql/workbench/parts/query/modelViewTab/queryModelViewTab.ts [hunk 3]
```diff
@@ -19,6 +20,19 @@ export class QueryModelViewTab implements IPanelTab {
...
 
+	public putState(dynamicModelViewTabsState: Map<string, QueryModelViewState>): void {
+		dynamicModelViewTabsState.set(this.view.componentId, this.view.state);
+	}
+
+	public captureState(dynamicModelViewTabsState: Map<string, QueryModelViewState>): void {
+		for (let i = 0; i < dynamicModelViewTabsState.keys.length; ++i) {
+			let currentIdentifier = dynamicModelViewTabsState[dynamicModelViewTabsState.keys[i]];
+			if (currentIdentifier === this.view.componentId) {
+				this.view.state = dynamicModelViewTabsState[dynamicModelViewTabsState.keys[i]];
```
Comment (kburtram) [inline src/sql/workbench/parts/query/modelViewTab/queryModelViewTab.ts]: should break the loop here to avoid unneeded iterations

**Pull request merged**